### PR TITLE
message(engine): waiting response is aware of client finished

### LINKED ISSUE #6639: Speed up QueryStatus for DM in engine
### Is your feature request related to a problem?

need 30s to complete this OpenAPI

```
retry 5 time(s), cmd:  curl "http://127.0.0.1:10245/api/v1/jobs/988936a3-55fd-45a9-acd7-7d024f908b49/status" | jq -e '.TaskStatus."mysql-01".Status.Stage == 4'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   485  100   485    0     0     16      0  0:00:30  0:00:30 --:--:--   128
```

and the log shows

```
3m3e-server-executor-0               | [2022/08/05 09:01:39.270 +00:00] [INFO] [worker.go:214] ["task stage changed"] [job_id=988936a3-55fd-45a9-acd7-7d024f908b49] [worker_id=9a5dce95-9317-4f5e-a569-984b8d2aa01c] [task-id=mysql-01] [from=2] [to=4]
3m3e-server-executor-0               | [2022/08/05 09:01:39.270 +00:00] [DEBUG] [message_agent.go:419] ["unregister topic"] [job_id=988936a3-55fd-45a9-acd7-7d024f908b49] [worker_id=9a5dce95-9317-4f5e-a569-984b8d2aa01c] [component=message-agent] [topic=DM/39383839333661332d353566642d343561392d616364372d376430323466393038623439/6d7973716c2d3031]
...
3m3e-server-master-0                 | [2022/08/05 09:01:39.648 +00:00] [DEBUG] [openapi.go:356] [debugging] [request="GET /api/v1/jobs/988936a3-55fd-45a9-acd7-7d024f908b49/status HTTP/1.1\r\nHost: 127.0.0.1:10245\r\nAccept: */*\r\nUser-Agent: curl/7.84.0\r\n\r\n"]
...
3m3e-server-executor-2               | [2022/08/05 09:01:39.650 +00:00] [DEBUG] [message_agent.go:285] ["send request"] [job_id=988936a3-55fd-45a9-acd7-7d024f908b49] [component=message-agent] [client-id=mysql-01] [command=QueryStatus] [req="{\"Task\":\"mysql-01\"}"]
3m3e-server-executor-2               | [2022/08/05 09:01:39.650 +00:00] [DEBUG] [client.go:267] ["Sending Message"] [topic=DM/39383839333661332d353566642d343561392d616364372d376430323466393038623439/6d7973716c2d3031] [seq=1]
```

### Describe the feature you'd like

When DM jobmaster knows the worker has become offline, itself should decide if it should break send message. For example, when the worker is the last one of worker sequence, no need to waiting for a response from it.

### Describe alternatives you've considered

_No response_

### Teachability, Documentation, Adoption, Migration Strategy

_No response_

### LINKED ISSUE #6687: Auto resume does not work properly
### What did you do?

```
3m3e-server-executor-1               | [2022/08/10 03:30:29.576 +00:00] [ERROR] [worker.go:275] ["task runs with error"] [job_id=c534cb96-7e94-41cf-bfff-30e96c5723ec] [worker_id=5381fa52-11b9-4fa1-8f3e-5bba269c7572] [task-id=mysql-01] ["error msg"="[{\"ErrCode\":50000,\"ErrClass\":\"not-set\",\"ErrScope\":\"not-set\",\"ErrLevel\":\"high\",\"Message\":\"[Lightning:Config:ErrInvalidConfig]unsupported `tikv-importer.on-duplicate` ()\"}]"]
3m3e-server-executor-1               | [2022/08/10 03:30:29.576 +00:00] [INFO] [worker.go:281] ["got auto resume strategy"] [job_id=c534cb96-7e94-41cf-bfff-30e96c5723ec] [worker_id=5381fa52-11b9-4fa1-8f3e-5bba269c7572] [task-id=mysql-01] [strategy="dispatch auto resume"]
...
3m3e-server-executor-1               | [2022/08/10 03:30:29.626 +00:00] [ERROR] [worker.go:275] ["task runs with error"] [job_id=c534cb96-7e94-41cf-bfff-30e96c5723ec] [worker_id=5381fa52-11b9-4fa1-8f3e-5bba269c7572] [task-id=mysql-01] ["error msg"="[{\"ErrCode\":50000,\"ErrClass\":\"not-set\",\"ErrScope\":\"not-set\",\"ErrLevel\":\"high\",\"Message\":\"[Lightning:Config:ErrInvalidConfig]unsupported `tikv-importer.on-duplicate` ()\"}]"]
3m3e-server-executor-1               | [2022/08/10 03:30:29.626 +00:00] [INFO] [worker.go:281] ["got auto resume strategy"] [job_id=c534cb96-7e94-41cf-bfff-30e96c5723ec] [worker_id=5381fa52-11b9-4fa1-8f3e-5bba269c7572] [task-id=mysql-01] [strategy="dispatch auto resume"]
```

### What did you expect to see?

should not "dispatch auto resume" too quickly

### What did you see instead?

see above

### Versions of the cluster

Dataflow Engine version (run `tiflow version`):

```console
(paste Dataflow Engine version here)
```

### PR DESCRIPTION
**Pull request by lance6716**
Signed-off-by: lance6716 <lance6716@gmail.com>

### What problem does this PR solve?

Issue Number: close #6639 close #6687

### What is changed and how it works?

add another context so waiting response can be aware of client finished

#### Tests 

 - Integration test
 - Manual test (add detailed scripts or steps below)

the response time is reduced from 30s to 2s

```
retry 5 time(s), cmd:  curl "http://127.0.0.1:10245/api/v1/jobs/1798b590-4b27-4841-8453-2ec35415c8f4/status" | jq -e '.TaskStatus."mysql-01".Status.Stage == 4'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   485  100   485    0     0    169      0  0:00:02  0:00:02 --:--:--   169
false
retry 6 time(s), cmd:  curl "http://127.0.0.1:10245/api/v1/jobs/1798b590-4b27-4841-8453-2ec35415c8f4/status" | jq -e '.TaskStatus."mysql-01".Status.Stage == 4'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   442  100   442    0     0  49165      0 --:--:-- --:--:-- --:--:-- 55250
true
```

### Release note 

```release-note
 `None`.
```

### CODE REVIEW

File: dm/config/task.go [hunk 1]
```diff
@@ -266,9 +266,10 @@ type LoaderConfig struct {
 // DefaultLoaderConfig return default loader config for task.
 func DefaultLoaderConfig() LoaderConfig {
 	return LoaderConfig{
-		PoolSize:   defaultPoolSize,
-		Dir:        defaultDir,
-		ImportMode: LoadModeSQL,
+		PoolSize:    defaultPoolSize,
+		Dir:         defaultDir,
+		ImportMode:  LoadModeSQL,
+		OnDuplicate: OnDuplicateReplace,
```
Comment (GMHDBJD) [inline dm/config/task.go:272]: Can we add a test case with minimal job config to test all default value? Maybe add in unit test like what we do in job_template

File: engine/pkg/dm/message_agent.go [hunk 2]
```diff
@@ -147,13 +156,25 @@ type MessageAgent interface {
...
+	// with this clientID should be treated as a different client.
+	RemoveClient(clientID string) error
 	SendMessage(ctx context.Context, clientID string, command string, msg interface{}) error
 	SendRequest(ctx context.Context, clientID string, command string, req interface{}) (interface{}, error)
 }
 
+type clientGroup struct {
+	mu sync.RWMutex
+	// key is client-id
+	m       map[string]client
```
Comment (amyangfei) [inline engine/pkg/dm/message_agent.go:173]: Is `m` short for map 🤔️

File: dm/config/task.go (hunk 1 above)
Comment (lance6716) [reply dm/config/task.go:272]: going to modify dm-job.yaml in current e2e test

**Review by amyangfei — APPROVED** [review]

File: engine/pkg/dm/message_agent.go (hunk 2 above)
Comment (lance6716) [reply engine/pkg/dm/message_agent.go:173]: yes. I'll rename it to `clients`

**Review by GMHDBJD — APPROVED** [review]
LGTM

File: engine/pkg/dm/message_agent.go [hunk 3]
```diff
@@ -147,13 +156,25 @@ type MessageAgent interface {
...
+	// When client offline temporary, caller should use this method with nil client.
 	UpdateClient(clientID string, client client) error
+	// RemoveClient is used when client is offline permanently, or the new client
+	// with this clientID should be treated as a different client.
+	RemoveClient(clientID string) error
 	SendMessage(ctx context.Context, clientID string, command string, msg interface{}) error
 	SendRequest(ctx context.Context, clientID string, command string, req interface{}) (interface{}, error)
 }
 
+type clientGroup struct {
```
Comment (GMHDBJD) [inline engine/pkg/dm/message_agent.go:170]: Shall we close all client context when message agent close?

File: engine/pkg/dm/message_agent.go [hunk 4]
```diff
@@ -248,6 +248,12 @@ func (agent *MessageAgentImpl) Close(ctx context.Context) error {
 		agent.cancel()
 	}
 	agent.wg.Wait()
+
```
Comment (GMHDBJD) [inline engine/pkg/dm/message_agent.go:251]: How about use agent.ctx instead of context.Backgroud() 😂

**Comment by lance6716** [comment]
/unhold

/merge

cc @GMHDBJD

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**